From a69e278e9a519455c6ca6f7522d805967a473637 Mon Sep 17 00:00:00 2001
From: Liam Howlett <Liam.Howlett@oracle.com>
Date: Thu, 24 Nov 2022 10:49:27 -0500
Subject: [PATCH 03/42] maple_tree: Detect dead nodes in ma_data_end()

If there is a dead node, then don't use any of the data and just return
0.  The caller will need to validate the node is not dead before using
the returned data anyways.  This is necessary for the RCU mode of the
maple tree.

Fixes: 54a611b60590 ("Maple Tree: add new data structure")
Signed-off-by: Liam Howlett <Liam.Howlett@oracle.com>
Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 lib/maple_tree.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/maple_tree.c b/lib/maple_tree.c
index cd19b615a..7f6c2b214 100644
--- a/lib/maple_tree.c
+++ b/lib/maple_tree.c
@@ -1409,6 +1409,9 @@ static inline unsigned char ma_data_end(struct maple_node *node,
 {
 	unsigned char offset;
 
+	if (unlikely(ma_dead_node(node)))
+		return 0;
+
 	if (type == maple_arange_64)
 		return ma_meta_end(node, type);
 
-- 
2.39.0.rc2.1.gbd5df96b79

